api_test.go

```diff
@@ -564,3 +565,75 @@ func TestGetOAuthClients(t *testing.T) {
 
 	ApiSpecRegister = nil
 }
+
+func TestResetHandler(t *testing.T) {
+	uri := "/tyk/reload/"
+
+	ApiSpecRegister = make(map[string]*APISpec)
+
+	makeSampleAPI(t, apiTestDef)
+
+	recorder := httptest.NewRecorder()
+	params := make(url.Values)
+
+	req, err := http.NewRequest("GET", uri+params.Encode(), nil)
+
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	resetHandler(recorder, req)
+
+	if recorder.Code != 200 {
+		t.Fatal("Hot reload failed, response code was: ", recorder.Code)
+	}
+
+	if len(ApiSpecRegister) == 0 {
+		t.Fatal("Hot reload was triggered but no APIs were found.")
+	}
+}
+
+func TestGroupResetHandler(t *testing.T) {
+	signalChan := make(chan bool)
+	cacheStore := RedisClusterStorageManager{}
+	cacheStore.Connect()
+
+	go func(pubSubChannel string, store RedisClusterStorageManager, signalChan chan bool) {
+		store.StartPubSubHandler(pubSubChannel, func(message redis.Message) {
+			notif := Notification{}
+			if err := json.Unmarshal(message.Data, &notif); err != nil {
+				t.Fatal("Unmarshalling message body failed, malformed: ", err)
+			}
+			if notif.Command == NoticeGroupReload {
+				signalChan <- true
```

Comment:
There definitely is a true or false here. This is the wrong way to write tests. This is a version of testing `time.Sleep` with a slightly larger timeout. Of course it will work 99.99% of the times when you run the test, but that's besides the point. The test takes longer than needed and it's flakey.

And as said above, any test using a timeout to work around introduced races will make https://github.com/TykTechnologies/tyk/issues/363 impossible.